File: 0328-26-3108-minimum-cost-walk-in-weighted-graph/0328-26-3108-minimum-cost-walk-in-weighted-graph.java
public class Solution {
    
    private int find(int u, int[] parent) {
        if (parent[u] != u) {
            parent[u] = find(parent[u], parent);
        }
        return parent[u];
    }
    
    private void join(int u, int v, int w, int[] parent, int[] costs) {
        int rootU = find(u, parent);
        int rootV = find(v, parent);
        costs[rootU] = costs[rootV] = (costs[rootU] & costs[rootV] & w);
        parent[rootV] = rootU;
    }
    
    private int calc(int u, int v, int[] parent, int[] costs) {
        int rootU = find(u, parent);
        int rootV = find(v, parent);
        if (rootU != rootV) {
            return -1;
        }
        return costs[rootU];
    }
    
    public int[] minimumCost(int n, int[][] edges, int[][] queries) {
        int[] parent = new int[n];
        int[] costs = new int[n];
        
        for (int i = 0; i < n; i++) {
            parent[i] = i;
            costs[i] = -1;
        }
        
        for (int[] edge : edges) {
            int u = edge[0];
            int v = edge[1];
            int w = edge[2];
            join(u, v, w, parent, costs);
        }
        
        int[] result = new int[queries.length];
        for (int i = 0; i < queries.length; i++) {
            int u = queries[i][0];
            int v = queries[i][1];
            result[i] = calc(u, v, parent, costs);
        }
        
        return result;
    }
}

File: 0328-26-3108-minimum-cost-walk-in-weighted-graph/0328-26-3108-minimum-cost-walk-in-weighted-graph.py
class Solution:
    def minimumCost(self, n: int, edges: List[List[int]], query: List[List[int]]) -> List[int]:
        parent = list(range(n))      
        costs = [-1]*n
        
        def find(u):
            if parent[u] != u:
                parent[u] = find(parent[u])
            return parent[u]

        def join(u,v,w):
            u = find(u)
            v = find(v)
            costs[u] = costs[v] = costs[u] & costs[v] & w
            parent[v] = u
            
        def calc(u,v):
            u = find(u)
            v = find(v)
            if u != v:
                return -1
            return costs[u]
        
        for u, v, w in edges:
            join(u, v, w)
            
        result = []
        
        for u, v in query:
            result.append(calc(u, v))
        return result

                
        
        